app/common/collections/runner.py

```diff
@@ -0,0 +1,179 @@
+
+    @classmethod
+    def load(cls, submission_id: UUID) -> "FormRunner":
+        return cls(SubmissionHelper.load(submission_id))
+
+    def context(
+        self, *, question_id: Optional[UUID] = None, form_id: Optional[UUID] = None, source: Optional[FormRunnerState]
+    ) -> "FormRunner":
+        if question_id and form_id:
+            raise ValueError("Expected only one of question_id or form_id")
+
+        self.source = source
+        self._tasklist_form = GenericSubmitForm()
+
+        if question_id:
+            self.question = self.submission.get_question(question_id)
+            self.form = self.question.form
+
+            context = self.submission.expression_context
+            self._question_form = build_question_form(self.question, context)(data=context)
+        if form_id:
+            self.form = self.submission.get_form(form_id)
+
+            all_questions_answered, _ = self.submission.get_all_questions_are_answered_for_form(self.form)
+            self._check_your_answers_form = CheckYourAnswersForm(
+                section_completed=(
+                    "yes" if self.submission.get_status_for_form(self.form) == SubmissionStatusEnum.COMPLETED else None
+                )
+            )
+            self._check_your_answers_form.set_is_required(all_questions_answered)
+
+        return self
+
+    @property
+    def question_form(self) -> "DynamicQuestionForm":
+        if not self.question:
+            raise RuntimeError("Question context not set")
+        return self._question_form
+
+    @property
+    def check_your_answers_form(self) -> "CheckYourAnswersForm":
+        if not self.form:
+            raise RuntimeError("Form context not set")
+        return self._check_your_answers_form
+
+    @property
+    def tasklist_form(self) -> "GenericSubmitForm":
+        return self._tasklist_form
+
+    def save_question_answer(self) -> None:
+        if not self.question:
+            raise RuntimeError("Question context not set")
+
+        self.submission.submit_answer_for_question(self.question.id, self.question_form)
+
+    def save_is_form_completed(self, user: "User") -> None:
+        if not self.form:
+            raise RuntimeError("Form context not set")
+
+        try:
+            self.submission.toggle_form_completed(
+                form=self.form,
+                user=user,
+                is_complete=self.check_your_answers_form.section_completed.data == "yes",
+            )
+        except ValueError as e:
+            self.check_your_answers_form.section_completed.errors.append(  # type:ignore[attr-defined]
+                "You must complete all questions before marking this section as complete"
+            )
+            raise ValueError("Failed to save form completed status") from e
+
+    def submit(self, user: "User") -> None:
+        try:
+            self.submission.submit(user)
+            notification_service.send_collection_submission(self.submission.submission)
```

Comment:
Interesting - maybe this means that we have to take the submit button out of the tasklist macro at some point? If we'd want to send the email in AGF but not in DGF for example.

Reply:
Ah very good point 👍 this was originally done at the view layer so maybe simply passing it on to the submission helper and having the view manage the email lifecycle - everything should be wrapped in the same transaction either way 

I'll make that tweak anticipating that I think it helps show where the delineation is